// ngx-autosize/bundles/ngx-autosize.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('rxjs'), require('rxjs/operators')) :
    typeof define === 'function' && define.amd ? define('ngx-autosize', ['exports', '@angular/core', 'rxjs', 'rxjs/operators'], factory) :
    (factory((global['ngx-autosize'] = {}),global.ng.core,global.rxjs,global.rxjs.operators));
}(this, (function (exports,core,rxjs,operators) { 'use strict';

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /** @type {?} */
    var MAX_LOOKUP_RETRIES = 3;
    var AutosizeDirective = /** @class */ (function () {
        function AutosizeDirective(element, _zone) {
            this.element = element;
            this._zone = _zone;
            this.onlyGrow = false;
            this.useImportant = false;
            this.retries = 0;
            this._destroyed$ = new rxjs.ReplaySubject(1);
            if (this.element.nativeElement.tagName !== 'TEXTAREA') {
                this._findNestedTextArea();
            }
            else {
                this.textAreaEl = this.element.nativeElement;
                this.textAreaEl.style.overflow = 'hidden';
                this._onTextAreaFound();
            }
        }
        /**
         * @param {?} textArea
         * @return {?}
         */
        AutosizeDirective.prototype.onInput = /**
         * @param {?} textArea
         * @return {?}
         */
            function (textArea) {
                this.adjust();
            };
        /**
         * @return {?}
         */
        AutosizeDirective.prototype.ngOnDestroy = /**
         * @return {?}
         */
            function () {
                this._destroyed$.next(true);
                this._destroyed$.complete();
            };
        /**
         * @return {?}
         */
        AutosizeDirective.prototype.ngAfterContentChecked = /**
         * @return {?}
         */
            function () {
                this.adjust();
            };
        /**
         * @param {?} changes
         * @return {?}
         */
        AutosizeDirective.prototype.ngOnChanges = /**
         * @param {?} changes
         * @return {?}
         */
            function (changes) {
                this.adjust(true);
            };
        /**
         * @return {?}
         */
        AutosizeDirective.prototype._findNestedTextArea = /**
         * @return {?}
         */
            function () {
                var _this = this;
                this.textAreaEl = this.element.nativeElement.querySelector('TEXTAREA');
                if (!this.textAreaEl && this.element.nativeElement.shadowRoot) {
                    this.textAreaEl = this.element.nativeElement.shadowRoot.querySelector('TEXTAREA');
                }
                if (!this.textAreaEl) {
                    if (this.retries >= MAX_LOOKUP_RETRIES) {
                        console.warn('ngx-autosize: textarea not found');
                    }
                    else {
                        this.retries++;
                        setTimeout(function () {
                            _this._findNestedTextArea();
                        }, 100);
                    }
                    return;
                }
                this.textAreaEl.style.overflow = 'hidden';
                this._onTextAreaFound();
            };
        /**
         * @return {?}
         */
        AutosizeDirective.prototype._onTextAreaFound = /**
         * @return {?}
         */
            function () {
                var _this = this;
                this._zone.runOutsideAngular(function () {
                    rxjs.fromEvent(window, 'resize')
                        .pipe(operators.takeUntil(_this._destroyed$), operators.debounceTime(200), operators.distinctUntilChanged())
                        .subscribe(function () {
                        _this._zone.run(function () {
                            _this.adjust();
                        });
                    });
                });
                setTimeout(function () {
                    _this.adjust();
                });
            };
        /**
         * @param {?=} inputsChanged
         * @return {?}
         */
        AutosizeDirective.prototype.adjust = /**
         * @param {?=} inputsChanged
         * @return {?}
         */
            function (inputsChanged) {
                if (inputsChanged === void 0) {
                    inputsChanged = false;
                }
                if (this.textAreaEl) {
                    /** @type {?} */
                    var currentText = this.textAreaEl.value;
                    if (inputsChanged === false &&
                        currentText === this._oldContent &&
                        this.textAreaEl.offsetWidth === this._oldWidth) {
                        return;
                    }
                    this._oldContent = currentText;
                    this._oldWidth = this.textAreaEl.offsetWidth;
                    /** @type {?} */
                    var clone = this.textAreaEl.cloneNode(true);
                    /** @type {?} */
                    var parent_1 = this.textAreaEl.parentNode;
                    clone.style.visibility = 'hidden';
                    parent_1.appendChild(clone);
                    clone.style.overflow = 'auto';
                    clone.style.height = 'auto';
                    /** @type {?} */
                    var height = clone.scrollHeight;
                    /** @type {?} */
                    var willGrow = height > this.textAreaEl.offsetHeight;
                    if (this.onlyGrow === false || willGrow) {
                        /** @type {?} */
                        var lineHeight = this._getLineHeight();
                        /** @type {?} */
                        var rowsCount = height / lineHeight;
                        /** @type {?} */
                        var styleAttribute = '';
                        if (this.minRows && this.minRows >= rowsCount) {
                            height = this.minRows * lineHeight;
                        }
                        else if (this.maxRows && this.maxRows <= rowsCount) {
                            height = this.maxRows * lineHeight;
                            styleAttribute += 'overflow: auto;';
                        }
                        else {
                            styleAttribute += 'overflow: hidden;';
                        }
                        styleAttribute += "height: " + height + "px";
                        styleAttribute += this.useImportant ? '!important;' : ';';
                        this.textAreaEl.setAttribute('style', styleAttribute);
                    }
                    parent_1.removeChild(clone);
                }
            };
        /**
         * @private
         * @return {?}
         */
        AutosizeDirective.prototype._getLineHeight = /**
         * @private
         * @return {?}
         */
            function () {
                /** @type {?} */
                var lineHeight = parseInt(this.textAreaEl.style.lineHeight, 10);
                if (isNaN(lineHeight) && window.getComputedStyle) {
                    /** @type {?} */
                    var styles = window.getComputedStyle(this.textAreaEl);
                    lineHeight = parseInt(styles.lineHeight, 10);
                }
                if (isNaN(lineHeight)) {
                    /** @type {?} */
                    var fontSize = window.getComputedStyle(this.textAreaEl, null).getPropertyValue('font-size');
                    lineHeight = Math.floor(parseInt(fontSize.replace('px', ''), 10) * 1.5);
                }
                return lineHeight;
            };
        AutosizeDirective.decorators = [
            { type: core.Directive, args: [{
                        selector: '[autosize]'
                    },] }
        ];
        /** @nocollapse */
        AutosizeDirective.ctorParameters = function () {
            return [
                { type: core.ElementRef },
                { type: core.NgZone }
            ];
        };
        AutosizeDirective.propDecorators = {
            minRows: [{ type: core.Input }],
            maxRows: [{ type: core.Input }],
            onlyGrow: [{ type: core.Input }],
            useImportant: [{ type: core.Input }],
            onInput: [{ type: core.HostListener, args: ['input', ['$event.target'],] }]
        };
        return AutosizeDirective;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var AutosizeModule = /** @class */ (function () {
        function AutosizeModule() {
        }
        AutosizeModule.decorators = [
            { type: core.NgModule, args: [{
                        declarations: [AutosizeDirective],
                        imports: [],
                        exports: [AutosizeDirective]
                    },] }
        ];
        return AutosizeModule;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */

    exports.AutosizeDirective = AutosizeDirective;
    exports.AutosizeModule = AutosizeModule;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
