// ngx-dropdown-list/esm2015/src/ngx-dropdown-list/filter/input-filter.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, EventEmitter, Input, Output } from '@angular/core';
export class InputFilterComponent {
    constructor() {
        /**
         * bind to [inputFilterBlur], emits with (blur) of filter input box
         */
        this.inputFilterBlur = new EventEmitter();
        /**
         * bind to [filterValueChange], for 2-way binding of filterValue
         */
        this.filterValueChange = new EventEmitter();
    }
    /**
     * triggers with (blur) event, emits the (inputFilterBlur) event
     * @param {?} event
     * @return {?}
     */
    onFilterTextBlur(event) {
        this.inputFilterBlur.emit(event);
    }
    /**
     * triggers with (input) event, emits the (filterValueChange) event for 2-way binding of filterValue
     * @return {?}
     */
    onChange() {
        this.filterValueChange.emit(this.filterValue);
    }
}
InputFilterComponent.decorators = [
    { type: Component, args: [{
                selector: 'input-filter',
                template: `
    <input type="text" class="filter-box" [(ngModel)]="filterValue" (input)="onChange()" (blur)="onFilterTextBlur($event)">
  `,
                styles: ["*,:after,:before{font-size:inherit;font-weight:inherit;font-family:inherit;box-sizing:inherit;background:inherit}.filter-box{width:calc(100% - 10px);height:28px;border-radius:4px;border:1px solid #ccc;margin:1px 5px 5px;padding-left:5px;font-size:12px;box-sizing:border-box;color:#495057}.filter-box:focus{outline:0;border-color:#ccc}"]
            }] }
];
InputFilterComponent.propDecorators = {
    filterValue: [{ type: Input }],
    inputFilterBlur: [{ type: Output }],
    filterValueChange: [{ type: Output }]
};
if (false) {
    /**
     * bind to [filterValue], the value of the filter
     * @type {?}
     */
    InputFilterComponent.prototype.filterValue;
    /**
     * bind to [inputFilterBlur], emits with (blur) of filter input box
     * @type {?}
     */
    InputFilterComponent.prototype.inputFilterBlur;
    /**
     * bind to [filterValueChange], for 2-way binding of filterValue
     * @type {?}
     */
    InputFilterComponent.prototype.filterValueChange;
}
