// ngx-dropdown-list/esm5/src/ngx-dropdown-list/dropdown-list.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FormsModule } from '@angular/forms';
import { ItemComponent } from './item/item.component';
import { GroupItemComponent } from './group-item/group-item.component';
import { InputFilterComponent } from './filter/input-filter.component';
import { AnchorComponent } from './anchor/anchor.component';
import { DropdownComponent } from './dropdown/dropdown.component';
import { DropdownListComponent } from './dropdown-list.component';
var DropdownListModule = /** @class */ (function () {
    function DropdownListModule() {
    }
    DropdownListModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [
                        DropdownListComponent,
                        ItemComponent,
                        GroupItemComponent,
                        InputFilterComponent,
                        AnchorComponent,
                        DropdownComponent
                    ],
                    exports: [DropdownListComponent],
                    imports: [CommonModule, FormsModule],
                    providers: [],
                    bootstrap: [],
                },] }
    ];
    return DropdownListModule;
}());
export { DropdownListModule };
