// ngx-dropdown-list/esm5/src/ngx-dropdown-list/item/item.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, EventEmitter, Input, Output } from '@angular/core';
import { isNumber } from '../utils/util';
var ItemComponent = /** @class */ (function () {
    function ItemComponent() {
        /**
         * bind to [formatNumber], the flag for formatting the number
         */
        this.formatNumber = false;
        /**
         * bind to [itemClick] event, triggers when clicking the item of dropdown
         */
        this.itemClick = new EventEmitter();
        /**
         * bind to [checkStatusChange] event, triggers when check status is changed in checkbox mode.
         */
        this.checkStatusChange = new EventEmitter();
    }
    /**
     * check whether needs to format number for the provided text
     */
    /**
     * check whether needs to format number for the provided text
     * @param {?} value
     * @return {?}
     */
    ItemComponent.prototype.needFormatNumber = /**
     * check whether needs to format number for the provided text
     * @param {?} value
     * @return {?}
     */
    function (value) {
        return isNumber(value) && this.formatNumber;
    };
    /**
     * triggered when clicking the item, emits the [itemClick] event
     */
    /**
     * triggered when clicking the item, emits the [itemClick] event
     * @param {?} item
     * @return {?}
     */
    ItemComponent.prototype.onItemClick = /**
     * triggered when clicking the item, emits the [itemClick] event
     * @param {?} item
     * @return {?}
     */
    function (item) {
        this.itemClick.emit(item);
    };
    /**
     * triggered when checking status changed in checkbox mode, emits the [checkStatusChange] event
     */
    /**
     * triggered when checking status changed in checkbox mode, emits the [checkStatusChange] event
     * @param {?} item
     * @return {?}
     */
    ItemComponent.prototype.onCheckStatusChange = /**
     * triggered when checking status changed in checkbox mode, emits the [checkStatusChange] event
     * @param {?} item
     * @return {?}
     */
    function (item) {
        this.checkStatusChange.emit(item);
    };
    ItemComponent.decorators = [
        { type: Component, args: [{
                    selector: 'item',
                    template: "\n    <label [class.container-checkbox]=\"checkbox\"\n           [class.container-selection]=\"!checkbox && !item.selected\"\n           [class.container-selection-selected]=\"!checkbox && item.selected\"\n           [id]=\"item.id\" (mousedown)=\"onItemClick(item)\">\n      {{needFormatNumber(item.text)? (item.text | number:'1.0-2') : item.text}}{{suffixText? suffixText : ''}}\n      <ng-container *ngIf=\"checkbox\">\n        <input type='checkbox' [id]=\"'checkbox-'+item.text\" (change)=\"onCheckStatusChange(item)\" [checked]=\"item.selected\">\n        <span class=\"checkmark\" [id]=\"'checkmark-'+item.text\"></span>\n      </ng-container>\n    </label>",
                    styles: ["*,:after,:before{font-size:inherit;font-weight:inherit;font-family:inherit;box-sizing:inherit;background:inherit}.container-checkbox,.container-selection,.container-selection-selected,.dropdown-item{background:0 0;display:list-item;list-style:none;position:relative;width:100%;height:auto;cursor:pointer;color:#495057;padding-bottom:5px;padding-top:5px;padding-left:12px}.container-checkbox{padding-left:35px}.container-checkbox input{position:absolute;opacity:0;cursor:pointer}.container-checkbox .checkmark{position:absolute;top:5px;left:10px;height:15px;width:15px;border:1px solid rgba(0,0,0,.3);background-color:#fff;border-radius:4px}.container-checkbox .checkmark:after{content:\"\";position:absolute;display:none;left:5px;top:2px;width:4px;height:7px;border:solid #fff;border-width:0 2px 2px 0;-webkit-transform:rotate(45deg);transform:rotate(45deg)}.container-checkbox:hover input~.checkmark{background-color:#ccc}.container-checkbox input:checked~.checkmark{background-color:#2196f3;border:1px solid #2196f3}.container-checkbox input:checked~.checkmark:after{display:block}.container-checkbox:hover{color:#66afe9}.container-selection,.container-selection-selected{padding-left:12px}.container-selection-selected:hover,.container-selection:hover{color:#495057;background:#e0ffff}.container-selection-selected{color:#fff;background:#6495ed}"]
                }] }
    ];
    ItemComponent.propDecorators = {
        checkbox: [{ type: Input }],
        item: [{ type: Input }],
        formatNumber: [{ type: Input }],
        suffixText: [{ type: Input }],
        itemClick: [{ type: Output }],
        checkStatusChange: [{ type: Output }]
    };
    return ItemComponent;
}());
export { ItemComponent };
if (false) {
    /**
     * bind to [checkbox], the flag of checkbox mode
     * @type {?}
     */
    ItemComponent.prototype.checkbox;
    /**
     * bind to [item], the dropdown option item
     * @type {?}
     */
    ItemComponent.prototype.item;
    /**
     * bind to [formatNumber], the flag for formatting the number
     * @type {?}
     */
    ItemComponent.prototype.formatNumber;
    /**
     * bind to [suffixText], the suffixText that will be displayed in the dropdown
     * @type {?}
     */
    ItemComponent.prototype.suffixText;
    /**
     * bind to [itemClick] event, triggers when clicking the item of dropdown
     * @type {?}
     */
    ItemComponent.prototype.itemClick;
    /**
     * bind to [checkStatusChange] event, triggers when check status is changed in checkbox mode.
     * @type {?}
     */
    ItemComponent.prototype.checkStatusChange;
}
